core/src/main/java/jenkins/security/ResourceDomainRootAction.java

```diff
@@ -0,0 +1,307 @@
+                        return;
+                    }
+                }
+            }
+
+            try (ACLContext ignored = ACL.as(auth)) {
+                try {
+                    Stapler.getCurrent().invoke(req, rsp, Jenkins.get(), requestUrlSuffix + restOfPath);
+                } catch (Exception ex) {
+                    // cf. UnwrapSecurityExceptionFilter
+                    Throwable cause = ex.getCause();
+                    while (cause != null) {
+                        if (cause instanceof AccessDeniedException) {
+                            throw (AccessDeniedException) cause;
+                        }
+                        cause = cause.getCause();
+                    }
+                    throw ex;
+                }
+                /*
+                While we could just redirect below to the real URL like we do for expired resource URLs, the question
+                is whether we'd end up in a redirect loop if the exception is specific to this mode (and the "normal"
+                URLs redirect to resource URLs). That seems even worse than an error here.
+                 */
+            } catch (AccessDeniedException ade) {
+                /* This is expected to be fairly common, as permission issues are thrown up as exceptions */
+                LOGGER.log(Level.FINE, "Failed permission check for resource URL access", ade);
+                rsp.sendError(403, "Failed permission check: " + ade.getMessage());
+            } catch (Exception e) {
+                /*
+                This should be fairly uncommon -- it's basically the 'rage' butler response. Notably, lack of access
+                to a job (permissions/deleted/renamed/...) would not throw an exception, but just sends a 404 response.
+                 */
+                LOGGER.log(Level.FINE, "Something else failed for resource URL access", e);
+                rsp.sendError(404);
+            }
+        }
+
+        @Override
+        public String toString() {
+            return "[" + super.toString() + ", authentication=" + authenticationName + "; key=" + browserUrl + "]";
+        }
+    }
+
+    public static class Token {
+        private String path;
+        private String username;
+        private Instant timestamp;
+        private Token (String path, @Nullable String username, Instant timestamp) {
+            this.path = path;
+            this.username = Util.fixNull(username);
+            this.timestamp = timestamp;
+        }
+
+        private String encode() {
+            String value = username + ":" + timestamp.toEpochMilli() + ":" + path;
+            byte[] valueBytes = value.getBytes(StandardCharsets.UTF_8);
+            byte[] byteValue = ArrayUtils.addAll(KEY.mac(valueBytes), valueBytes);
+            return Base64.getUrlEncoder().encodeToString(byteValue);
+        }
+
+        private static Token decode(String value) {
+            byte[] byteValue = Base64.getUrlDecoder().decode(value);
```

Comment:
For reference, this needs to be in the try block:
```java
	at java.util.Base64$Decoder.decode0(Base64.java:734)
	at java.util.Base64$Decoder.decode(Base64.java:526)
	at java.util.Base64$Decoder.decode(Base64.java:549)
	at jenkins.security.ResourceDomainRootAction$Token.decode(ResourceDomainRootAction.java:280)
	at jenkins.security.ResourceDomainRootAction$Token.access$000(ResourceDomainRootAction.java:262)
	at jenkins.security.ResourceDomainRootAction.getDynamic(ResourceDomainRootAction.java:111)
	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
```

Reply:
Done in #4302